src/module_wrap.cc

```diff
@@ -559,14 +564,27 @@ static MaybeLocal<Promise> ImportModuleDynamically(
     THROW_ERR_EXECUTION_ENVIRONMENT_NOT_AVAILABLE(isolate);
     return MaybeLocal<Promise>();
   }
+  Realm* realm = Realm::GetCurrent(context);
+  if (realm == nullptr) {
+    // Fallback to the principal realm if it's in a vm context.
+    realm = env->principal_realm();
+  }
 
   EscapableHandleScope handle_scope(isolate);
 
   Local<Function> import_callback =
-    env->host_import_module_dynamically_callback();
+      realm->host_import_module_dynamically_callback();
 
   Local<FixedArray> options = host_defined_options.As<FixedArray>();
-  if (options->Length() != HostDefinedOptions::kLength) {
+  // When the dynamic import is initiated by a realm top-level request, e.g.
+  // ShadowRealm.prototype.importValue, the host defined options should be
+  // empty.
+  bool is_realm_top_level_import =
```

Comment:
Notice that the bug reported in https://github.com/nodejs/node/issues/49726 can also trigger this even when it does not come from a shadow realm.

Reply:
Actually I wonder if we can extend the HostImportModuleDynamically callback to explicitly pass this information in the arguments...the length alone doesn't seem to be a very good reference here.

Reply:
Actually, I wonder why we don't get a referrer in this case...[the spec](https://tc39.es/proposal-shadowrealm/#sec-shadowrealmimportvalue) says:

> 6. Perform [HostImportModuleDynamically](https://tc39.es/ecma262/#sec-hostimportmoduledynamically)(null, specifierString, innerCapability).

I wonder why the referrer needs to be null here instead of e.g. the `[[ShadowRealm]]` slot of the shadow realm object. Is is by design that the host gets no information about the referrer here? That seems a bit odd..

Reply:
This is different from the eval case.

`ShadowRealm.prototype.importValue` is a built-in API and should avoid resolving the module specifiers by determining who called it, which would be a case of dynamic scope that we should avoid.

As for propagating active script in V8, I would suggest waiting for discussion at https://github.com/tc39/ecma262/pull/3195.

Reply:
> There is currently an edge case in V8 such that some code that makes no use of shadow realms can result in an empty referrer
> Which means the edge case now would fail DCHECK_EQ(realm->kind(), Realm::kShadowRealm) which doesn't seem quite right...

I believe this case would not be able to reach the DCHECK because there is an `is_realm_top_level_import && realm->kind() != Realm::kShadowRealm` branch to disallow empty host-defined options in non-shadow-realm realms.

However, this is true for `Promise.resolve('import(`./mod`)').then(eval)` in ShadowRealm, which would not throw a `TypeError: Invalid host defined options` but the module resolution would proceed. Indeed, this can be an inconsistency between the ShadowRealms and principal realms.

I'll submit a separate PR to add a fallback for empty host-defined options in all realms.